l3kernel/l3sort.dtx

```diff
@@ -968,31 +1021,641 @@
+      {
+        \@@_quick_split:NnNn #2 \@@_quick_end:nnTFNn { } \s_@@_mark
+          {
+            \@@_quick_split:NnNn #1
+              \@@_quick_end:nnTFNn { } \s_@@_mark {#5}
+            {#3}
+          }
+        #6 \s_@@_stop
+      }
+  }
+%    \end{macrocode}
+% \end{macro}
+%
+%    \begin{macrocode}
+\cs_if_exist:NT \tex_expanded:D
   {
-    \@@_quick_split:NnNn #1
-      \@@_quick_end:nnTFNn { } \s_@@_mark {#5}
-    {#3}
-    #6 \s_@@_stop
+%    \end{macrocode}
+%
+% \begin{macro}[EXP]
+%   {\@@_if_stop:w, \@@_if_mark:w, \@@_if_tail:w}
+%   Since the three scan marks are forbidden to appear in the lists we can setup
+%   quick tests by just gobbling everything up until that marker.
+%    \begin{macrocode}
+    \cs_new:Npn \@@_if_stop:w #1 \s_@@_merge_exp_stop {}
+    \cs_new:Npn \@@_if_mark:w #1 \s_@@_merge_exp_mark {}
+    \cs_new:Npn \@@_if_tail:w #1 \s_@@_merge_exp_tail {}
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}[rEXP]
+%   {
+%     \@@_merge_exp_bottom_up_tl:Nnn,
+%     \@@_merge_exp_bottom_up_tl_mark:w
+%     \@@_merge_exp_bottom_up_tl_stop:w
+%   }
+%   The first bottom up pass grabs two items and sorts them. Markers are placed
+%   to indicate the end of each sublist. Since we grab two items in each
+%   recursion step we have to check both of these values for the end of list
+%   markers. If the input token list has an odd number of elements, |#3| will be
+%   \cs{s_@@_merge_exp_mark} in the last recursion. In this case |#2| will be
+%   the last element and form a sublist with just one element. If the input
+%   token list has an even number of elements, |#3| will be
+%   \cs{s_@@_merge_exp_stop} in the last recursion and we just leave the markers
+%   for the end of sublists.
+%
+%   After the first bottom up pass every sublist will be sorted and followed by
+%   \cs{s_@@_merge_exp_tail}. The sublists are separated by
+%   \cs{s_@@_merge_exp_list}, and the end of sublists is indicated by a sublist
+%   containing only \cs{s_@@_merge_exp_mark} and another one containing
+%   \cs{s_@@_merge_exp_stop}.
+%    \begin{macrocode}
+    \cs_new:Npn \@@_merge_exp_bottom_up_tl:Nnn #1#2#3
+      {
+        \@@_if_stop:w #3 \@@_merge_exp_bottom_up_tl_stop:w \s_@@_merge_exp_stop
+        \@@_if_mark:w #3 \@@_merge_exp_bottom_up_tl_mark:w \s_@@_merge_exp_mark
+        #1 {#2} {#3}
+          { \__kernel_exp_not:w { {#2} {#3} } }
+          { \__kernel_exp_not:w { {#3} {#2} } }
+        \s_@@_merge_exp_tail \s_@@_merge_exp_list
+        \@@_merge_exp_bottom_up_tl:Nnn #1
+      }
+    \cs_new:Npn \@@_merge_exp_bottom_up_tl_stop:w
+        \s_@@_merge_exp_stop
+        \@@_if_mark:w \s_@@_merge_exp_stop \@@_merge_exp_bottom_up_tl_mark:w
+        \s_@@_merge_exp_mark
+        #1
+        \s_@@_merge_exp_tail \s_@@_merge_exp_list
+        \@@_merge_exp_bottom_up_tl:Nnn #2
+      {
+        \s_@@_merge_exp_mark \s_@@_merge_exp_list
+        \s_@@_merge_exp_stop \s_@@_merge_exp_list
+      }
```

Comment:
I think this level of detail in the delimiters of macro parameters is not readable.  Strictly speaking you are right that it is faster for TeX, but I would still go for the slightly slower `\cs_new:Npn \@@_merge_exp_bottom_up_tl_stop:w #1 \s_@@_merge_exp_list #2#3`.  Maybe I'm wrong though.